### ISSUE TITLE
Open-Source Release Status

### ISSUE THREAD

**Issue by arthur-flam** [Samsung/qaboard#1]
Go!

Welcome to QA-Board! We are working on open-sourcing the project and are targeting an *official* release for Q3 2020.

## What needs to be done before the release
- [x] Pre-release, as all items marked 🌱 are done.

### Open-source process
- [x] Authorization to release as OSS 🌱
- [x] Licence (Apache 2) 🌱

### Docs
- [x] Docs: Getting Started. (Available at https://samsung.github.io/qaboard/) 🌱
- [x] Website
- [x] Docs with common use-cases
- [x] README
- [x] Server bring-up guide

### Technical changes
- [x] `qa init` creates a functional project
- [x] Merge the `qatools` and `qaboard` internal repositories 🌱
- [x] Make sure nothing is hardcoded for our infra : pass secrets at least as ENV 🌱variables or config files, and handle mount naming convention with a centralized config.... 🌱
- [x] Decide on a way to pick a default location where to store the output data. Needs to be mounted by the container and writable easily.
- [x] Ensure 1min server bringup with e.g. `docker-compose`
- [x] Remove `USER arthurf` from *Dockerfile*, maybe make it configurable to avoid `NFS+squash_root` issues.

### API changes:
- [x] `qaboard-backend.slamvizapp>backend.backend`, `qaboard-webapp>webapp`
- [x] *qatools.yaml->qaboard.yaml*
- [x] Import as `qaboard` not qatools
- [x] Output folder naming conventions with hashes
- [x] Friendlier `run()` API (`ctx.obj`->`ctx` : RunContext)
- [x]  `sed -i s/qatools/qaboard/g *`

### Features
- [x] Make it easy to support more async task queues - [Wiki](https://github.com/Samsung/qaboard/wiki/adding-new-runners)
- [x] Support for at least one common task queue beyond what we have internally at Samsung, and explain how to implement support for  others.
- [x] Create a public CI
- [x] Support more git servers (e.g. Github). Don't even require Read rights.
- [x] Make it easier to debug visualizations and metrics with per-batch settings

### PR
- [x] Public demo server
- [x] All good for a release blog post

**Comment by amirfru** [comment]
Hey @arthur-flam ! 
Can I ask which task-runners is it tested for? thanks!

**Comment by arthur-flam** [comment]
Hi Amir ! Even better, please suggest something you are using!

I planned on adding support for something easy to setup and flexible like `celery` or `python-rq`.

For other integrations (unless there is a big demand) I'll write docs, make sure the code is easy to extend, and rely on help from the community.

**Comment by arthur-flam** [comment]
This week I improved the docs a lot, they are now good enough to be published. I expect to have them online early next week, along with a website.

I also published the CLI wrapper to pypi:
```bash
pip install qaboard
```
> As soon as I merge the two repositories composing QAboard today, I'll push the code online

**Comment by arthur-flam** [comment]
Hi, some updates:
- Merged the two repos that composed `qaboard`
- Added basic support for "secrets" to pass e.g. gitlab credentials.

We're getting closer to a pre-release. What's missing is creating a "master-open-source" branch with a clean history, replacing a few default / docs, and pushing it..!

**Comment by arthur-flam** [comment]
Update:
- there is now a fully sanitized version of the code. I'll publish it ASAP now 😌
- I uploaded the website/docs to the `gh-pages` branch, but GitHub Pages isn't setup correctly so we'll have to wait for it to be live...

**Comment by arthur-flam** [comment]
The code is online 🎉🎉🎉 !
There is still work to do, but things feel more concrete now. 🎆

**Comment by arthur-flam** [comment]
The website is live https://samsung.github.io/qaboard !

**Comment by arthur-flam** [comment]
Hi!
I've been making progress refactoring batches to allow more types of runners. For now the API I picked is
```python
class MyRunner():
  type: str  # local, celery, ray, kubernetes...
  start(blocking=True)
  stop()

  @static
  start_jobs(jobs, job_options)
  @static
  stop_jobs(jobs, job_options)

  # those function would use
  self.RunContext.job_options
  # to know the selected concurrency, settings to connect to the task queue...
```
@amirfru hi! Any update about what you'd need?

I expect to push the branch in the next few days, possibly directly merge on `master`.

**Comment by arthur-flam** [comment]
✨ There is now support for Celery ! https://samsung.github.io/qaboard/docs/celery-integration/
Tuning from the web application is still hardcoded to use our infra... But that will be fixed soon.

**Comment by arthur-flam** [comment]
Hi! I quick update to say that we're still working on the open-source release, although we've focused more on quality of life features and small fixes this past month. Stay tuned!

**Comment by arthur-flam** [comment]
Hello!
Last month we've been busy with other priorities, still some notable improvements landed:
- support for ["matrix" batches](https://samsung.github.io/qaboard/docs/batches-running-on-multiple-inputs/#matrix-batches)
- filtering is now much more performant
- lossless image viewer support

**Comment by arthur-flam** [comment]
Hi!
I am finishing adding support for `docker-compose`:
- the *whole* application stack can already be started in 1 command, and there is a structured way to specify custom ENV variables.
- makes it easier to implement a public CI, host a demo server, or just have users try QA-Board.
- will help us provide pre-built container images.
- will provide an easy migration path to more complex orchestration tools (e.g. k8s..).
- Other benefits: the app now has only seconds of downtime during upgrades, is served from a unique port, no need for cron jobs managed separately, and our internal tweaks are much more isolated.

I expect to push the changes and new docs in the coming days. Bye!

**Comment by arthur-flam** [comment]
**We now use `docker-compose`!**
- New [guide to start a QA-Board server](https://samsung.github.io/qaboard/docs/start-server)
- Faster startup/update time, as all services are split
- Everything is served from a single port, we reviewed the whole mess of reverse proxies.
- The docs are available from the application at `/docs`
- Almost all the code is the same in the OSS and our internal version.
- We will shortly add in _docker-compose.yml_  a sample cron job doing backups

We also started using **github actions to create a public CI:**
- the website is already updated automatically from the docs
- we're building the docker images. We need to add some code to check and push them to github's docker registry. We might also upload the CLI package to pypy.

> The next task is not requiring read-access to the git repositories of the users' projects - or at least support github in addition to gitlab (currently required...)

**Comment by arthur-flam** [comment]
Hi!

Here are some updates:
- QA-Board now doesn't require Gitlab anymore, nor read-right to projects' repositories. Using Gitlab enables nice-to-have features like showing user avatars, but we'll work on doing it with Github too.
- Each batch can now have its own config and metrics, making using with `qa --share` much more pleasant.
- If you've missed it, we now support a new visualization, [flame graphs](https://github.com/Samsung/qaboard/issues/11#issuecomment-646220354).

**We're now in new territory: anyone can start using QA-board!** We'll check it's as smooth as promised by the docs, and give updates.

**Comment by arthur-flam** [comment]
There is now a demo server at [demo.qamarks.com](https://demo.qamarks.com/arthurf/locator/commit/1b790e58b367d7e857d0f64d4114cce272883e1e?batch=%40ec2-user%7C%20default&selected_views=output-list&reference=1b790e58b367d7e857d0f64d4114cce272883e1e&batch_ref=%40ec2-user%7C%20default&filter_ref=%3A2&filter=%3A1)!
- It was straightforward to setup just following the guide, so we can now feel at ease looking for users.
- For now there is a single demo project comparing 2 versions of a map-based algorithm. We'll add more demos and polish it...

**Comment by arthur-flam** [comment]
Some major updates:

1. The API was simplified:
  - Instead of the ugly `context.obj["configurations"]`, there is `context.configs`. And so on.
  - Tuning comes for free, no need to worry anymore about `context.obj["extra_parameters"]`
  - For convenience there is now `context.params`
  - The docs were updated and _changes are backward compatible_.

2. Users can now save artifacts and outputs at different locations, and it's all a bit simpler to specify. Go to [the docs to read more](https://samsung.github.io/qaboard/docs/storage/where-is-the-data).

> Those changes close the TODOs to make QA-Board "public-ready" 🚀 !
>
> **I'm closing the issue!**
